src/wbase/SendChannelShared.h

```diff
@@ -47,8 +50,32 @@ class TransmitLock;
 namespace wbase {
 
 /// A class that provides a SendChannel object with synchronization so it can be
-/// shared by across multiple threads. Due to what may be sent, the synchronization locking
-/// is needs to be available outside of the class.
+/// shared by across multiple threads. Due to what may be sent, the synchronization
+/// locking needs to be available outside of the class.
```

Comment:
I have a question regarding "synchronization locking outside of the class" `wbase::SendChannelShared`. How does this ("external synchronization") work with the class that still has the private atomics, one private condition variable, and three private mutexes?
```
    std::mutex _streamMutex;
    std::mutex _transmitLockMtx;
    std::condition_variable _transmitLockCv;
    std::atomic<bool> _schemaColsSet{false};
    std::atomic<uint32_t> _scsSeq{0};
    std::atomic<bool> _firstChannelSqlConn{true};
    std::atomic<bool> _schemaColsSet{false};
    mutable std::mutex _tMtx;
    ...
```
I'm just concerned that debugging the run-time behavior of the class would be very hard.

I have an impression that considering many (8) synchronization primitives, the class represents "multiple faces" (roles, etc.). Is there any way to restructure the class into something safer?

Reply:
It looks like the synchronization outside of the class is no longer necessary, but I want to be very careful about removing it. _streamMutex is the only member that could be accessed externally.

Reply:
I've removed any need to externally lock _streamMutex. The other mutexes protect different things, which is essential to prevent deadlock and for performance reasons. It is probably possible to refactor the code so there's only one or two mutex per class, but its complicated enough that it should go in another ticket.